### PR TITLE
Feature/extensions list

### PR DESCRIPTION
**Pull request by NothingEverHappens**
image

### CODE REVIEW

File: src/components/Extensions/api/useExtensionRowSpec.tsx [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+      ...backend.extensionVersion.spec,
+      // TODO(kirjs): Use default icon for local extensions
+      iconUri: backend.extension.iconUri,
+      publisherIconUri: backend.extension.publisher?.iconUri,
+    };
+  });
+}
+
+export function useExtensionRowSpecs(): ExtensionRowSpec[] {
+  const backends = useExtensionBackends();
```
Comment (tjlav5) [inline src/components/Extensions/api/useExtensionRowSpec.tsx:40]: Why not useExtensions?

File: src/components/Extensions/api/useExtensionBackends.tsx [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+  backend: Backend
+): backend is ExtensionBackend {
+  return (
+    !!backend.extensionInstanceId &&
+    ((!!backend.extension && !!backend.extensionVersion) ||
+      !!backend.extensionSpec)
+  );
+}
+
+export function useExtensionBackends(): ExtensionBackend[] {
```
Comment (tjlav5) [inline src/components/Extensions/api/useExtensionBackends.tsx:32]: Isn't this UseExtensions?

File: src/components/Extensions/api/useExtensionRowSpec.tsx [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+import { ExtensionRowSpec } from '../models';
+import { useExtensionBackends } from './useExtensionBackends';
+import { ExtensionBackend } from './useExtensions';
+
+export function convertBackendsToRawSpecs(backends: ExtensionBackend[]) {
+  return backends.map((backend) => {
+    if ('extensionSpec' in backend) {
+      return {
+        id: backend.extensionInstanceId,
+        ...backend.extensionSpec,
```
Comment (tjlav5) [inline src/components/Extensions/api/useExtensionRowSpec.tsx:25]: Why are we flattening this?

File: src/components/Extensions/api/useExtensionRowSpec.tsx (hunk 1 above)
Comment (tjlav5) [reply src/components/Extensions/api/useExtensionRowSpec.tsx:40]: I think a useExtensionSpec hook makes sense instead?

File: src/components/Extensions/index.tsx [hunk 4]
```diff
@@ -46,39 +49,16 @@ export interface Backend {
...
 const HydrateExtensions: React.FC = ({ children }) => {
   const backends = useExtensionBackends();
+  const specs = useExtensionRowSpecs();
 
   return (
-    <ExtensionsProvider extensionBackends={backends}>
-      {children}
-    </ExtensionsProvider>
+    <ExtensionsSpecProvider extensionRowSpecs={specs}>
+      <ExtensionsProvider extensionBackends={backends}>
```
Comment (tjlav5) [inline src/components/Extensions/index.tsx:58]: I'm a little confused here. I'd imagine a list of extension-backends at this level, but then at the list/details-level there would be hooks to get the scoped Spec

**Review by tjlav5 — CHANGES_REQUESTED** [review]
UI looks great, just come commentary on the API changes

File: src/components/Extensions/api/useExtensionRowSpec.tsx (hunk 3 above)
Comment (NothingEverHappens) [reply src/components/Extensions/api/useExtensionRowSpec.tsx:25]: Resolving after discussing offline

**Comment by NothingEverHappens** [comment]
PTAL

File: src/components/Extensions/api/internal/useExtensionData.test.tsx [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+
+import { renderHook } from '@testing-library/react-hooks';
+import React, { Suspense } from 'react';
+
+import { TestEmulatorConfigProvider } from '../../../common/EmulatorConfigProvider';
+import { mockExtensionBackends } from '../../testing/mockExtensionBackend';
+import { BACKEND_LIST, CONFIG_WITH_EXTENSION } from '../../testing/utils';
+import { useExtensionData } from './useExtensionData';
+
+describe('useExtensionDatas', () => {
```
Comment (tjlav5) [inline src/components/Extensions/api/internal/useExtensionData.test.tsx:25]:
```suggestion
describe('useExtensionData', () => {
```

File: src/components/Extensions/api/internal/useExtensionData.tsx [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+import { useExtensionBackends } from './useExtensionBackends';
+
+export function useExtensionData(): Extension[] {
+  const backends = useExtensionBackends();
+
+  return convertBackendsToExtensions(backends);
+}
+
+export function convertBackendToExtension(backend: ExtensionBackend) {
+  if ('extensionSpec' in backend) {
```
Comment (tjlav5) [inline src/components/Extensions/api/internal/useExtensionData.tsx:27]: Do we still have the isLocalExtension type-guard?

File: src/components/Extensions/api/internal/useExtensionData.tsx [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import { Extension } from '../../models';
+import { ExtensionBackend } from '../useExtensions';
+import { useExtensionBackends } from './useExtensionBackends';
+
+export function useExtensionData(): Extension[] {
```
Comment (tjlav5) [inline src/components/Extensions/api/internal/useExtensionData.tsx:20]: nit: useExtensionsData?

File: src/components/Extensions/models.ts [hunk 8]
```diff
@@ -136,6 +136,12 @@ export interface ExtensionSpec {
   externalServices?: ExternalService[];
 }
 
+export interface Extension extends ExtensionSpec {
```
Comment (tjlav5) [inline src/components/Extensions/models.ts:139]: One more time for the record... I'm not super excited about extending here, I'd prefer composition - but don't consider this a blocker.

**Review by tjlav5 — CHANGES_REQUESTED** [review]

File: src/components/Extensions/models.ts (hunk 8 above)
Comment (NothingEverHappens) [reply src/components/Extensions/models.ts:139]: I overall support, but want to do it a bit later, once I have better understanding of things

**Pull request merged**